### ISSUE TITLE
Clients do not handle gen_tcp:send/2 errors

### ISSUE THREAD

**Issue by dreverri** [basho/riak_repl#69]
The `riak_repl_tcp_client` does not handle the return from `gen_tcp:send/2` which may return `{error, Reason}`. This can cause replication clients to hang around indefinitely without replication actually occurring.

A module has been provided below to help observe current client behavior. To observe the issue you will need to setup 2 Riak nodes and connect them via riak-repl. Once that is setup, you can use the below module as follows:

``` erlang
riak_repl_client_test:test_client_ports().                                 
[{<0.8179.0>,ok},{<0.8182.0>,ok}]

riak_repl_client_test:close_client_ports().                                  
[{<0.8179.0>,true},{<0.8182.0>,true}]

riak_repl_client_test:test_client_ports(). 
[{<0.8179.0>,{error,closed}},{<0.8182.0>,{error,closed}}]

riak_repl_client_test:test_client_ports_through_client().
[{<0.8179.0>,
  {status,[{node,'node1@127.0.0.1'},
           {site,"node2-to-node1"},
           {strategy,riak_repl_keylist_client},
           {fullsync_worker,<0.8184.0>},
           {put_pool_size,5},
           {connected,"127.0.0.1",9012},
           {state,wait_for_fullsync}]}},
 {<0.8182.0>,
  {status,[{node,'node1@127.0.0.1'},
           {site,"node3-to-node1"},
           {strategy,riak_repl_keylist_client},
           {fullsync_worker,<0.8186.0>},
           {put_pool_size,5},
           {connected,"127.0.0.1",9013},
           {state,wait_for_fullsync}]}}]
```

The last command, `riak_repl_client_test:test_client_ports_through_client/0`, forces the client to send a `keapalive_ack` over the closed port. The expectation is that this would cause the client to fail and re-connect. The client does not fail. The client ignores the `{error, closed}` returned from `gen_tcp:send/2`.

``` erlang
-module(riak_repl_client_test).

-compile(export_all).

test_client_ports_through_client() ->
    Pids = get_client_pids(),
    [{Pid, send_keepalive_through_client(Pid)} || Pid <- Pids].

test_client_ports() ->
    Ports = get_client_ports(),
    [{Pid, send_keepalive(Port)} || {Pid, Port} <- Ports].

close_client_ports() ->
    Ports = get_client_ports(),
    [{Pid, close_port(Port)} || {Pid, Port} <- Ports].

get_client_ports() ->
    Pids = get_client_pids(),
    [{Pid, get_port_from_pid(Pid)} || Pid <- Pids].

get_client_pids() ->
   [ Pid2 || {riak_repl_client_sup,P,_,_} <- supervisor:which_children(riak_repl_sup),
           {_,Pid2,_,_} <- supervisor:which_children(P)].

get_port_from_pid(Pid) ->
    get_port_from_status(sys:get_status(Pid)).

get_port_from_status(Status) ->
    {_, _, _, [_,running,_,_,[_,_,{data,[{_,State}]}]]} = Status,
    element(5, State).

close_port(Port) ->
    erlang:port_close(Port).

send_keepalive(Port) ->
    gen_tcp:send(Port, term_to_binary(keepalive)).

send_keepalive_through_client(Pid) ->
    Port = get_port_from_pid(Pid),
    Pid ! {tcp, Port, term_to_binary(keepalive)},
    riak_repl_tcp_client:status(Pid).
```

**Comment by Vagabond** [comment]
Does the gh56-zombie-servers branch help with this at all?

**Comment by dreverri** [comment]
Yes, I believe that patch would solve this issue.

**Comment by bookshelfdave** [comment]
Verified that the client reconnects with PR 102 while running the test above.

**Comment by Vagabond** [comment]
Closed by #102.
